README.md
# Introduction

Check password strength against several rules. Includes ActiveRecord/ActiveModel support.

<a href="https://travis-ci.org/fnando/password_strength"><img src="https://travis-ci.org/fnando/password_strength.svg" alt="Build Status" /></a>

Validates the strength of a password according to several rules:

* size
* 3+ numbers
* 2+ special characters
* uppercased and downcased letters
* combination of numbers, letters and symbols
* password contains username
* sequences (123, abc, aaa)
* repetitions
* can't be a common password (view list at support/common.txt)

Some results:

* `123`: weak
* `123abc`: weak
* `aaaaaa`: weak
* `myPass145`: good
* `myPass145$`: strong

## Install

```
gem install password_strength
```

or put this in your Gemfile:

```ruby
gem "password_strength"
```

The JavaScript code is also available as a NPM package.

```
npm install @fnando/password_strength --save
```

If you want the source go to http://github.com/fnando/password_strength

## Usage

```ruby
strength = PasswordStrength.test("johndoe", "mypass")
#=> return a object

strength.good?
#=> status == :good

strength.weak?
#=> status == :weak

strength.strong?
#=> status == :strong

strength.status
#=> can be :weak, :good, :strong

strength.valid?(:strong)
#=> strength == :strong

strength.valid?(:good)
#=> strength == :good or strength == :strong
```

## ActiveRecord/ActiveModel

The PasswordStrength library comes with ActiveRecord/ActiveModel support.

```ruby
class Person < ActiveRecord::Base
  validates_strength_of :password
end
```

To be honest, you can use it with plain ActiveModel objects.

```ruby
class Person
  include ActiveModel::Model
  validates_strength_of :password
end

# or simply

class Person
  include ActiveModel::Validations
  validates_strength_of :password
end
```

The default options are `:level => :good, :with => :username`.

If you want to compare your password against other field, you have to set the `:with` option.

```ruby
validates_strength_of :password, :with => :email
```

The available levels are: `:weak`, `:good` and `:strong`.

```ruby
validates_strength_of :password, :with => :email, :level => :good
```

Also you can set level with a lambda.

```ruby
validates_strength_of :password, :with => :email, :level => lambda {|u| :good }
```

You can also provide a custom class/module that will test that password.

```ruby
validates_strength_of :password, :using => CustomPasswordTester
```

Your `CustomPasswordTester` class should override the default implementation. In practice, you're going to override only the `test` method that must call one of the following methods: `invalid!`, `weak!`, `good!` or `strong!`.

```ruby
class CustomPasswordTester < PasswordStrength::Base
  def test
    if password != "mypass"
      invalid!
    else
      strong!
    end
  end
end
```

The tester above will accept only +mypass+ as password.

PasswordStrength implements two validators: `PasswordStrength::Base` and `PasswordStrength::Validators::Windows2008`.

**ATTENTION:** Custom validators are not supported by JavaScript yet!

## JavaScript

The PasswordStrength also implements the algorithm in the JavaScript.

```javascript
var strength = PasswordStrength.test("johndoe", "mypass");
strength.isGood();
strength.isStrong();
strength.isWeak();
strength.isValid("good");
```

The API is basically the same!

You can use the `:exclude` option. Only regular expressions are supported for now.

```javascript
var strength = PasswordStrength.test("johndoe", "password with whitespaces", {exclude: /\s/});
strength.isInvalid();
```

Additionaly, a jQuery plugin is available.

```javascript
$.strength("#username", "#password");
```

The line above will validate the `#password` field against `#username`.
The result will be an image to the respective strength status. By default the image path will be
`/images/weak.png`, `/images/good.png` and `/images/strong.png`.

You can overwrite the image path and the default callback.

```javascript
$.strength.weakImage = "/weak.png";
$.strength.goodImage = "/good.png";
$.strength.strongImage = "/strong.png";
$.strength.callback = function(username, password, strength) {
    // do whatever you want
};
```

If you just want to overwrite the callback, you can simple do

```javascript
$.strength("#username", "#password", function(username, password, strength){
    // do whatever you want
});
```

Get the files:

* https://github.com/fnando/password_strength/blob/master/app/assets/javascripts/jquery_strength.js
* https://github.com/fnando/password_strength/blob/master/app/assets/javascripts/password_strength.js

If you're using asset pipeline, just add the following lines to your `application.js`.

```javascript
//= require jquery
//= require password_strength
//= require jquery_strength
```

## Running tests

### Ruby

#### For current version
1. Install all dependencies with `bundle install`.
2. Run `rake test`.

#### Execute tests with appraisal gem
For generating a version gemfile for all gem versions configured on aprraisal gem, you have to execute:
`bundle exec appraisal install`

You can execute tests with diferentes gemfiles.lock. Example:

*Gemfile with activemodel 5.2.3*
`bundle exec appraisal activemodel-5.2.3 rake test`

You can check all created gemfiles for differents versions on _gemfiles_ folder.

### JavaScript

1. Install Node.js, then run `npm install`.
2. Open `test/password_strength_test.html` in your target browser.

## License

(The MIT License)

Copyright © 2010-2016 Nando Vieira (http://nandovieira.com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the ‘Software’), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED ‘AS IS’, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.0] - 2019-08-02
### Added
- Add support for Rails 5
- Add support for Ruby 2.3.3

lib/password_strength/base.rb
module PasswordStrength
  class Base
    MULTIPLE_NUMBERS_RE = /\d.*?\d.*?\d/
    MULTIPLE_SYMBOLS_RE = /[!@#\$%^&*?_~-].*?[!@#\$%^&*?_~-]/
    SYMBOL_RE = /[!@#\$%^&*?_~-]/
    UPPERCASE_LOWERCASE_RE = /([a-z].*[A-Z])|([A-Z].*[a-z])/
    LOWERCASE_RE = /[a-z]/
    UPPERCASE_RE = /[A-Z]/
    INVALID = :invalid
    WEAK = :weak
    STRONG = :strong
    GOOD = :good

    # Hold the username that will be matched against password.
    attr_accessor :username

    # The password that will be tested.
    attr_accessor :password

    # The score for the latest test. Will be +nil+ if the password has not been tested.
    attr_reader   :score

    # The current test status. Can be +:weak+, +:good+, +:strong+ or +:invalid+.
    attr_reader   :status

    # Set what characters cannot be present on password.
    # Can be a regular expression or array.
    #
    #   strength = PasswordStrength.test("john", "password with whitespaces", :exclude => [" ", "asdf"])
    #   strength = PasswordStrength.test("john", "password with whitespaces", :exclude => /\s/)
    #
    # Then you can check the test result:
    #
    #   strength.valid?(:weak)
    #   #=> false
    #
    #   strength.status
    #   #=> :invalid
    #
    attr_accessor :exclude

    def initialize(username, password, options = {})
      @username = username.to_s
      @password = password.to_s
      @score = 0
      @exclude = options[:exclude]
    end

    # Check if the password has the specified score.
    # Level can be +:weak+, +:good+ or +:strong+.
    def valid?(level = GOOD)
      case level
      when STRONG then
        strong?
      when GOOD then
        good? || strong?
      else
        !invalid?
      end
    end

    # Check if the password has been detected as strong.
    def strong?
      status == STRONG
    end

    # Mark password as strong.
    def strong!
      @status = STRONG
    end

    # Check if the password has been detected as weak.
    def weak?
      status == WEAK
    end

    # Mark password as weak.
    def weak!
      @status = WEAK
    end

    # Check if the password has been detected as good.
    def good?
      status == GOOD
    end

    # Mark password as good.
    def good!
      @status = GOOD
    end

    # Check if password has invalid characters based on PasswordStrength::Base#exclude.
    def invalid?
      status == INVALID
    end

    # Mark password as invalid.
    def invalid!
      @status = INVALID
    end

    # Return the score for the specified rule.
    # Available rules:
    #
    # * :password_size
    # * :numbers
    # * :symbols
    # * :uppercase_lowercase
    # * :numbers_chars
    # * :numbers_symbols
    # * :symbols_chars
    # * :only_chars
    # * :only_numbers
    # * :username
    # * :sequences
    def score_for(name)
      score = 0

      case name
      when :password_size then
        if password.size < 6
          score = -100
        else
          score = password.size * 4
        end
      when :numbers then
        score = 5 if password =~ MULTIPLE_NUMBERS_RE
      when :symbols then
        score = 5 if password =~ MULTIPLE_SYMBOLS_RE
      when :uppercase_lowercase then
        score = 10 if password =~ UPPERCASE_LOWERCASE_RE
      when :lowercase then
        score = 10 if password =~ LOWERCASE_RE
      when :uppercase then
        score = 10 if password =~ UPPERCASE_RE
      when :numbers_chars then
        score = 15 if password =~ /[a-z]/i && password =~ /[0-9]/
      when :numbers_symbols then
        score = 15 if password =~ /[0-9]/ && password =~ SYMBOL_RE
      when :symbols_chars then
        score = 15 if password =~ /[a-z]/i && password =~ SYMBOL_RE
      when :only_chars then
        score = -15 if password =~ /^[a-z]+$/i
      when :only_numbers then
        score = -15 if password =~ /^\d+$/
      when :username then
        if password == username
          score = -100
        else
          score = -15 if password =~ /#{Regexp.escape(username)}/
        end
      when :sequences then
        score = -15 * sequences(password)
        score += -15 * sequences(password.to_s.reverse)
      when :repetitions then
        score += -(repetitions(password, 2) * 4)
        score += -(repetitions(password, 3) * 3)
        score += -(repetitions(password, 4) * 2)
      end

      score
    end

    # Run all tests on password and return the final score.
    def test
      @score = 0

      if contain_invalid_matches?
        invalid!
      else
        @score += score_for(:password_size)
        @score += score_for(:numbers)
        @score += score_for(:symbols)
        @score += score_for(:uppercase_lowercase)
        @score += score_for(:lowercase)
        @score += score_for(:uppercase)
        @score += score_for(:numbers_chars)
        @score += score_for(:numbers_symbols)
        @score += score_for(:symbols_chars)
        @score += score_for(:only_chars)
        @score += score_for(:only_numbers)
        @score += score_for(:username)
        @score += score_for(:sequences)
        @score += score_for(:repetitions)

        @score = 0 if score < 0
        @score = 100 if score > 100

        weak!   if score < 35
        good!   if score >= 35 && score < 70
        strong! if score >= 70
      end

      score
    end

    def contain_invalid_matches? # :nodoc:
      return false unless exclude
      regex = exclude
      regex = /#{exclude.collect {|i| Regexp.escape(i)}.join("|")}/ if exclude.kind_of?(Array)
      password.to_s =~ regex
    end

    def repetitions(text, size) # :nodoc:
      text = text.mb_chars
      count = 0
      matches = []

      0.upto(text.size - 1) do |i|
        substring = text[i, size]

        next if matches.include?(substring) || substring.size < size

        matches << substring
        occurrences = text.scan(/#{Regexp.escape(substring)}/).length
        count += 1 if occurrences > 1
      end

      count
    end

    def sequences(text) # :nodoc:
      matches = 0
      sequence_size = 0
      bytes = []

      text.to_s.each_byte do |byte|
        previous_byte = bytes.last
        bytes << byte

        if previous_byte && ((byte == previous_byte + 1) || (previous_byte == byte))
          sequence_size += 1
        else
          sequence_size = 0
        end

        matches += 1 if sequence_size == 2
      end

      matches
    end
  end
end


javascripts/password_strength.js
var PasswordStrength = function() {
	var MULTIPLE_NUMBERS_RE = /\d.*?\d.*?\d/;
	var MULTIPLE_SYMBOLS_RE = /[!@#$%^&*?_~].*?[!@#$%^&*?_~]/;
	var UPPERCASE_LOWERCASE_RE = /([a-z].*[A-Z])|([A-Z].*[a-z])/;
	var LOWERCASE_RE = /[a-z]/;
	var UPPERCASE_RE = /[A-Z]/;
	var SYMBOL_RE = /[!@#\$%^&*?_~]/;

	this.username = null;
	this.password = null;
	this.score = 0;
	this.status = null;

	this.test = function() {
		this.score = 0;

		if (this.containInvalidMatches()) {
			this.status = "invalid";
		} else {
			this.score += this.scoreFor("password_size");
			this.score += this.scoreFor("numbers");
			this.score += this.scoreFor("symbols");
			this.score += this.scoreFor("uppercase_lowercase");
			this.score += this.scoreFor("uppercase");
			this.score += this.scoreFor("lowercase");
			this.score += this.scoreFor("numbers_chars");
			this.score += this.scoreFor("numbers_symbols");
			this.score += this.scoreFor("symbols_chars");
			this.score += this.scoreFor("only_chars");
			this.score += this.scoreFor("only_numbers");
			this.score += this.scoreFor("username");
			this.score += this.scoreFor("sequences");
			this.score += this.scoreFor("repetitions");

			if (this.score < 0) {
				this.score = 0;
			}

			if (this.score > 100) {
				this.score = 100;
			}

			if (this.score < 35) {
				this.status = "weak";
			}

			if (this.score >= 35 && this.score < 70) {
				this.status = "good";
			}

			if (this.score >= 70) {
				this.status = "strong";
			}
		}

		return this.score;
	};

	this.scoreFor = function(name) {
		score = 0;

		switch (name) {
			case "password_size":
				if (this.password.length < 4) {
					score = -100;
				} else {
					score = this.password.length * 4;
				}
				break;

			case "numbers":
				if (this.password.match(MULTIPLE_NUMBERS_RE)) {
					score = 5;
				}
				break;

			case "symbols":
				if (this.password.match(MULTIPLE_SYMBOLS_RE)) {
					score = 5;
				}
				break;

			case "uppercase_lowercase":
				if (this.password.match(UPPERCASE_LOWERCASE_RE)) {
					score = 10;
				}
				break;
			case "uppercase":
				if (this.password.match(UPPERCASE_RE)) {
					score = 10;
				}
				break;
			case "lowercase":
				if (this.password.match(LOWERCASE_RE)) {
					score = 10;
				}
				break;

			case "numbers_chars":
				if (this.password.match(/[a-z]/i) && this.password.match(/[0-9]/)) {
					score = 15;
				}
				break;

			case "numbers_symbols":
				if (this.password.match(/[0-9]/) && this.password.match(SYMBOL_RE)) {
					score = 15;
				}
				break;

			case "symbols_chars":
				if (this.password.match(/[a-z]/i) && this.password.match(SYMBOL_RE)) {
					score = 15;
				}
				break;

			case "only_chars":
				if (this.password.match(/^[a-z]+$/i)) {
					score = -15;
				}
				break;

			case "only_numbers":
				if (this.password.match(/^\d+$/i)) {
					score = -15;
				}
				break;

			case "username":
				if (this.password == this.username) {
					score = -100;
				} else if (this.password.indexOf(this.username) != -1) {
					score = -15;
				}
				break;

			case "sequences":
				score += -15 * this.sequences(this.password);
				score += -15 * this.sequences(this.reversed(this.password));
				break;

			case "repetitions":
				score += -(this.repetitions(this.password, 2) * 4);
        		score += -(this.repetitions(this.password, 3) * 3);
        		score += -(this.repetitions(this.password, 4) * 2);
				break;
		};

		return score;
	};

	this.isGood = function() {
		return this.status == "good";
	};

	this.isWeak = function() {
		return this.status == "weak";
	};

	this.isStrong = function() {
		return this.status == "strong";
	};

	this.isInvalid = function() {
	  return this.status == "invalid";
	};

	this.isValid = function(level) {
		if(level == "strong") {
			return this.isStrong();
		} else if (level == "good") {
			return this.isStrong() || this.isGood();
		} else {
			return !this.containInvalidMatches();
		}
	};

	this.containInvalidMatches = function() {
		if (!this.exclude) {
			return false;
		}

		if (!this.exclude.test) {
			return false;
		}

		return this.exclude.test(this.password.toString());
	};

	this.sequences = function(text) {
		var matches = 0;
		var sequenceSize = 0;
		var codes = [];
		var len = text.length;
		var previousCode, currentCode;

		for (var i = 0; i < len; i++) {
			currentCode = text.charCodeAt(i);
			previousCode = codes[codes.length - 1];
			codes.push(currentCode);

			if (previousCode) {
				if (currentCode == previousCode + 1 || previousCode == currentCode) {
					sequenceSize += 1;
				} else {
					sequenceSize = 0;
				}
			}

			if (sequenceSize == 2) {
				matches += 1;
			}
		}

		return matches;
	};

	this.repetitions = function(text, size) {
		var count = 0;
  		var matches = {};
		var len = text.length;
		var substring;
		var occurrences;
		var tmpText;

		for (var i = 0; i < len; i++) {
			substring = text.substr(i, size);
			occurrences = 0;
			tmpText = text;

			if (matches[substring] || substring.length < size) {
				continue;
			}

			matches[substring] = true;

			while ((i = tmpText.indexOf(substring)) != -1) {
				occurrences += 1;
				tmpText = tmpText.substr(i + 1);
			};

			if (occurrences > 1) {
				count += 1;
			}
		}

		return count;
	};

	this.reversed = function(text) {
		var newText = "";
		var len = text.length;

		for (var i = len -1; i >= 0; i--) {
			newText += text.charAt(i);
		}

		return newText;
	};
};

PasswordStrength.test = function(username, password) {
	strength = new PasswordStrength();
	strength.username = username;
	strength.password = password;
	strength.test();
	return strength;
};
